lib/afinter.c

```diff
@@ -635,3 +641,107 @@ afinter_get_metrics(void)
 {
   return metrics;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  AFInterSource *self;
+
+  if (is_live_collection)
+    return AFINTER_LIVE_COLLECTION_RUNNING;
+
+  if (current_internal_source != NULL)
+    return AFINTER_INTERNAL_SRC_PRESENT;
+
+  self = g_new0(AFInterSource, 1);
+  AFInterSourceOptions *options = g_new0(AFInterSourceOptions, 1);
+
+  afinter_source_options_defaults(options);
+  self->options = options;
+  stats_counter_set(metrics.queue_capacity, self->options->queue_capacity);
+
+  g_mutex_lock(&internal_msg_lock);
+  current_internal_source = self;
+  g_mutex_unlock(&internal_msg_lock);
+
+  is_live_collection = TRUE;
+
+  return AFINTER_LIVE_COLLECTION_STARTED;
+}
+
+void
+afinter_reset_live_collection(void)
+{
+  afinter_stop_live_collection();
+
+  if (is_live_collection)
+    {
+      if (internal_msg_queue)
+        _release_internal_msg_queue();
+    }
+}
+
+void
+afinter_stop_live_collection(void)
+{
+  if (is_live_collection)
+    {
+      if (current_internal_source != NULL)
+        {
+          AFInterSource *self = current_internal_source;
+
+          g_mutex_lock(&internal_msg_lock);
+          current_internal_source = NULL;
+          g_mutex_unlock(&internal_msg_lock);
+
+          is_live_collection = FALSE;
+          g_free(self);
+        }
+    }
+}
+
+void
+afinter_get_size_of_internal_logs(GString *result)
+{
+  g_string_append_printf(result, "Size of internal logs: %d", g_queue_get_length(internal_msg_queue));
+}
+
+void
+afinter_get_collected_messages(GString *result)
+{
+  LogMessage *log_msg;
+  GString *msg;
+  LogTemplate *template = NULL;
+
```

Comment:
this one is also needs a better concurrency protection

you must be sure that e.g. the `internal_msg_queue` stat is not changing between the NULL check and the pop

Reply:
@HofiOne : okay, I added protection for the whole block.